docs/SubmittingChanges.md

```diff
@@ -0,0 +1,159 @@
+> enhancement, separate those changes into two commits. If your change includes
+> a cleanup that extracts common functionality into a new function and a
+> bug fix in that common functionality, separate those into two commits.
+>
+> On the other hand, if you make a single change to numerous files, such as
+> adopting a new API provided by LLVM, group those changes into a single
+> commit.  Thus a single logical change is contained within a single commit.
+>
+> The point to remember is that each commit should make an easily understood
+> change that can be verified by reviewers. Each commit should be justifiable
+> on its own merits.
+
+Each commit must be a strict improvement. When submitting a series of commits,
+make a reasonable effort to ensure that there are no regressions in the middle
+of the series. This allows us to use `git bisect` when regressions do occur.
+
+Exercise judgment as to whether a series of commits should be submitted as
+multiple pull requests or as a single one. GitHub makes it difficult to review
+individual commits in a pull request, but having many pull requests is
+confusing for reviewers as well, because their relationship, if any, is not
+clear in the UI. Some rules of thumb:
+
+* If your changes are entirely separate, such as a new feature and an unrelated
+  bug fix that you happened to make while developing the feature, put the
+  commits onto separate branches and submit them as separate pull requests.
+* If your changes are related, such as a sequence of refactorings and
+  enhancements towards implementing a larger feature, consider submitting them
+  as a single pull requests. Reviewers can still look at individual commits
+  via GitHub's UI or by downloading your branch and examining it locally.
+
+Learn about useful tools in Git that help build series of logical commits:
+
+* `git gui` to interactively commit only a subset of local changes, and to
+  amend the most recent commit.
+* `git add -p` to achieve the same via the command line.
+* `git commit --amend` to amend the most recent commit from the command line.
+* `git stash` to temporarily undo a change and re-apply it later.
+* `git commit --squash=<commit>` and `git commit --fixup=<commit>` to mark
+  changes to be added to an earlier commit.
+* `git rebase -i <commit>` and the `rebase.autoSquash` option to rework a
+  series of commits.
+* `git rebase -i <commit> -x "..."` to execute a command after each commit;
+  you can use this to verify that LLPC compiles and passes lit tests after
+  each commit.
+* `git cherry-pick` for moving a commit from one branch to another.
+* `gitk` and `tig` to browse the repository's history.
+* `git reflog` to find old states of the repository that may seem lost; once
+  you commit something, Git will **not** lose it -- at least not for the next
+  30 days.
+
+Similar functionality is available in some GUI tools. Learning the tools of the
+trade makes you a better developer.
```

Comment:
I have mixed feeling about including this brief git tutorial in the document. The most important llpc-specific thing here is that we attempt to maintain linear history, everything else seems a bit out of scope of this document to me.
Wouldn't it be better to keep this doc shorter and instead reference some existing git tutorial? I'd expect everybody to either be familiar with git or inexperienced enough to get lost in this explanation.

Reply:
It wasn't meant as a tutorial so much as a list of conceptual links, to help with "not knowing what you don't know" type of situations. Linking to a Git tutorial doesn't necessarily achieve this goal, unless you happen to know of a specific tutorial that would? (I don't, unfortunately.) But I can see how it doesn't really fit here, so I'm moving it into its own section below.